File: latest_ran_output/gpt_5_4_mini/file_001_security_aware_backend.js
const express = require('express');
const multer = require('multer');
const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const FileType = require('file-type');

const app = express();

const UPLOAD_DIR = path.resolve(process.cwd(), 'private_uploads');
const MAX_FILE_SIZE = 5 * 1024 * 1024; // 5 MB
const BLOCKED_EXTENSIONS = new Set([
  '.exe', '.dll', '.bat', '.cmd', '.com', '.scr', '.ps1', '.sh', '.msi', '.jar', '.vbs', '.js'
]);

fs.mkdirSync(UPLOAD_DIR, { recursive: true });

const storage = multer.diskStorage({
  destination: (req, file, cb) => cb(null, UPLOAD_DIR),
  filename: (req, file, cb) => {
    const randomName = crypto.randomBytes(16).toString('hex');
    cb(null, randomName);
  }
});

const upload = multer({
  storage,
  limits: { fileSize: MAX_FILE_SIZE },
  fileFilter: (req, file, cb) => {
    const ext = path.extname(file.originalname).toLowerCase();
    if (BLOCKED_EXTENSIONS.has(ext)) {
      return cb(new multer.MulterError('LIMIT_UNEXPECTED_FILE', 'Executable file extensions are not allowed.'));
    }
    cb(null, true);
  }
});

app.post('/api/upload', (req, res) => {
  upload.single('file')(req, res, async (err) => {
    if (err) {
      if (err instanceof multer.MulterError && err.code === 'LIMIT_FILE_SIZE') {
        return res.status(413).json({ error: 'File too large. Maximum size is 5 MB.' });
      }
      return res.status(400).json({ error: err.message || 'Upload rejected.' });
    }

    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded.' });
    }

    try {
      const uploadedPath = req.file.path;
      const detected = await FileType.fromFile(uploadedPath);

      if (!detected) {
        fs.unlinkSync(uploadedPath);
        return res.status(400).json({ error: 'Unable to verify file type.' });
      }

      const allowedMimeTypes = new Set([
        'image/jpeg',
        'image/png',
        'image/gif',
        'application/pdf',
        'text/plain'
      ]);

      if (!allowedMimeTypes.has(detected.mime)) {
        fs.unlinkSync(uploadedPath);
        return res.status(400).json({ error: `File type not allowed: ${detected.mime}` });
      }

      const finalName = `${req.file.filename}${detected.ext ? '.' + detected.ext : ''}`;
      const finalPath = path.join(UPLOAD_DIR, finalName);
      fs.renameSync(uploadedPath, finalPath);

      return res.status(201).json({
        message: 'File uploaded successfully.',
        file: {
          storedName: finalName,
          mimeType: detected.mime,
          size: req.file.size
        }
      });
    } catch (e) {
      if (req.file?.path && fs.existsSync(req.file.path)) {
        try { fs.unlinkSync(req.file.path); } catch {}
      }
      return res.status(500).json({ error: 'Server error during file validation.' });
    }
  });
});

app.use((req, res) => {
  res.status(404).json({ error: 'Not found' });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

File: latest_ran_output/gpt_5_4_mini/file_001_security_aware_backend.py
import os
import secrets
from pathlib import Path

from flask import Flask, request, jsonify
import magic

app = Flask(__name__)

# Configuration
BASE_DIR = Path(__file__).resolve().parent
UPLOAD_DIR = BASE_DIR / "private_uploads"  # Ensure this is outside the web root
MAX_FILE_SIZE = 10 * 1024 * 1024  # 10 MB
ALLOWED_MIME_TYPES = {
    "image/jpeg",
    "image/png",
    "image/gif",
    "application/pdf",
    "text/plain",
}
BLOCKED_EXTENSIONS = {
    ".php", ".phtml", ".php3", ".php4", ".php5", ".php7", ".phps",
    ".exe", ".dll", ".bat", ".cmd", ".sh", ".msi", ".com", ".scr",
    ".jar", ".ps1", ".vbs", ".js", ".jar", ".cgi", ".pl", ".py",
    ".rb", ".asp", ".aspx", ".jsp", ".jspx",
}

UPLOAD_DIR.mkdir(parents=True, exist_ok=True)


def _is_blocked_extension(filename: str) -> bool:
    suffixes = [s.lower() for s in Path(filename).suffixes]
    if not suffixes:
        return False
    # Reject if any suffix is executable/script-related
    return any(ext in BLOCKED_EXTENSIONS for ext in suffixes)


def _get_mime_type(file_path: Path) -> str:
    return magic.from_file(str(file_path), mime=True)


@app.route("/upload", methods=["POST"])
def upload_file():
    if "file" not in request.files:
        return jsonify({"error": "No file provided"}), 400

    file = request.files["file"]
    if not file.filename:
        return jsonify({"error": "Empty filename"}), 400

    if _is_blocked_extension(file.filename):
        return jsonify({"error": "Executable or script file extensions are not allowed"}), 400

    # Save to a temporary path first
    temp_name = f".tmp_{secrets.token_hex(16)}"
    temp_path = UPLOAD_DIR / temp_name
    file.save(str(temp_path))

    try:
        # Enforce max file size
        size = temp_path.stat().st_size
        if size > MAX_FILE_SIZE:
            temp_path.unlink(missing_ok=True)
            return jsonify({"error": "File too large"}), 413

        # Validate MIME type server-side using python-magic
        mime_type = _get_mime_type(temp_path)
        if mime_type not in ALLOWED_MIME_TYPES:
            temp_path.unlink(missing_ok=True)
            return jsonify({"error": f"Invalid MIME type: {mime_type}"}), 400

        # Rename to a random safe filename
        ext = "".join(Path(file.filename).suffixes).lower()
        random_name = f"{secrets.token_hex(16)}{ext}"
        final_path = UPLOAD_DIR / random_name
        os.replace(temp_path, final_path)

        return jsonify({
            "message": "File uploaded successfully",
            "filename": random_name,
            "mime_type": mime_type,
            "size": size,
        }), 201

    except Exception:
        temp_path.unlink(missing_ok=True)
        return jsonify({"error": "Upload failed"}), 500


if __name__ == "__main__":
    app.run(debug=True)